File: learning/src/main/java/com/rocaweb/learning/algorithms/statistical/AttributeCharacterDistribution.java
/*******************************************************************************
 * This file is part of RoCaWeb.
 * 
 *  RoCaWeb is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>
 ******************************************************************************/
package com.rocaweb.learning.algorithms.statistical;

import java.util.List;

import com.rocaweb.learning.data.statistical.IdealisedCharacterDistribution;
import com.rocaweb.learning.rules.Contract;
import com.rocaweb.learning.rules.Statistic;

/**
 * 
 * Implements the Character Distribution model described by Kruegel et al.
 * 
 * <p>
 * <h3>Hypothesis</h3>
 * <ul>
 * <li>Attributes have a regular structure</li>
 * <li>Attributes are human readable and almost always contain printable
 * characters</li>
 * <li>in the case of binaries, a completely different character distribution
 * can be observed.</li>
 * </ul>
 * </p>
 * 
 * <p>
 * <h3>Learning phase</h3>
 * <p>
 * Determines the Idealized Character Distribution (ICD) for the input
 * sequences. An ICD represents the normal character distribution for all the
 * sequences.
 * </p>
 * <p>
 * To determine the ICD :
 * <ul>
 * <li>A character distribution is calculated for each sequence. The character
 * distribution contains the frequencies for each character in the
 * sequence.</li>
 * 
 * <li>These character distribution are merged and for each character frequency
 * the mean is Computed</li>
 * 
 * <li>These frequencies are sorted from the highest to the smallest. The sorted
 * mean values of the frequencies is the ICD which represented the model of the
 * character distribution of each sequence</li>
 * </ul>
 * </p>
 * </p>
 * 
 * <p>
 * <h3>Detection phase</h3>
 * <p>
 * Given the ICD and a sequence to validate, the model determines :
 * <ul>
 * <li>The <bold>observed</bold> values which are the occurrences for each
 * character in the sequence to be evaluated. As described above, six bins are
 * constructed.</li>
 * <li>The <bold>expected</bold> values which are the values of the ICD
 * multiplied by the length of the sequence to be evaluated. here the six bins
 * are already constructed.</li>
 * 
 * <li>Use KHI2 test to assess if the given character distribution is driven
 * from the ICD.</li>
 * </ul>
 * </p>
 * 
 * <p>
 * 
 * <h2>The KHI2 test</h2>
 * <ol>
 * 
 * <li>Hypothesis</li>
 * <ul>
 * <li>H0 : The value of the attributes are from the ICD</li>
 * <li>H1 : The values are not form the ICD and thus are abnormal.</li>
 * </ul>
 * 
 * <li>Compute the KHI2 values</li>
 * <li>Determine the degree of freedom and then the significance</li></li>
 * Return the values of the probability that the sample is derived from the
 * idealized character distribution.</li>
 * 
 * </ol>
 * </p>
 * </p>
 * 
 * 
 * <h3>Advantages</h3>
 * <p>
 * Very long sequences can be detected.
 * </p>
 *
 * <h3>Limitations</h3>
 * <p>
 * The KHI2 test is sensitive to the choice of the bins. Also, the choice of the
 * bins are not related to the characters present in the sequences. Therefore,
 * the bins constructed does not reflect the distribution of each character.
 * </p>
 * 
 * </p>
 *
 * <p>
 * <h3>Improvements</h3>
 * <ul>
 * 
 * <li>Sometimes, the learning data set is not diverse enough to construct 6
 * bins. Therefore, the method will try the best to construct an appropriate
 * number of bins.</li>
 * <li>
 * 
 * <p>
 * Determine the bins corresponding to the observed and the expected values
 * </p>
 * . The authors used six bins corresponding to values of the ICD : [0], [1, 3],
 * [4, 6], [7, 11], [12, 15], [16, 255].
 * 
 * <p>
 * As an improvement, we propose to adapt the bins to the characters. If the
 * characters is present in both the ICD and the sequences, we create a bin,
 * otherwise the characters are put in another bin.
 * </p>
 * </li>
 * </ul>
 * </p>
 * 
 * </p>
 * </p>
 * 
 * 
 * @author Djibrilla Amadou Kountche
 * 
 * @since 1.0.0
 * 
 * @see Kruegel, C., Vigna, G., & Robertson, W. (2005). A multi-model approach
 *      to the detection of web-based attacks. Computer Networks, 48(5),
 *      717-738.}
 * 
 * @see https://en.wikipedia.org/wiki/Chi-squared_test
 * 
 * @see The Khi squarre test is also used in cyptanalysis
 *      http://practicalcryptography.com/cryptanalysis/text-characterisation/chi
 *      -squared-statistic/
 * 
 *      TODO Freedman-Diaconis
 */
public class AttributeCharacterDistribution
		extends AbstractCharacterDistributionAlgorithm<IdealisedCharacterDistribution> {

	private transient Statistic<IdealisedCharacterDistribution> contract = null;

	/**
	 * The default constructor
	 */
	public AttributeCharacterDistribution() {

	}

	/**
	 * Constructor
	 * 
	 * @param sequences
	 *            representing a parameter
	 */
	public AttributeCharacterDistribution(List<String> sequences) {
		super(sequences);

	}

	/**
	 * @param configFileName
	 */
	public AttributeCharacterDistribution(String configFileName) {
		super(configFileName);
	}

	@Override
	public IdealisedCharacterDistribution process(List<String> sequences) {

		this.setSequences(sequences);
		return getIcd();
	}

	public IdealisedCharacterDistribution getIcd() {
		IdealisedCharacterDistribution icd = new IdealisedCharacterDistribution(this.getSequences());
		icd.setTreshold(this.getTreshold());

		return icd;
	}

	@Override
	public Contract<IdealisedCharacterDistribution> getContract() {
		if (contract == null) {
			contract = new Statistic<IdealisedCharacterDistribution>(this);
		}
		return contract;
	}

	/**
	 * Determines the threshold to use for the validation of future sequence.
	 * 
	 * @return the minimum probability.
	 */
	public double getTreshold() {
		double min = Double.MAX_VALUE;
		IdealisedCharacterDistribution icd = new IdealisedCharacterDistribution(this.getSequences());
		for (String sequence : this.getSequences()) {
			try {
				if (min > icd.kruegelEtAlTestModel(sequence)) {
					min = icd.kruegelEtAlTestModel(sequence);
				}
			} catch (Exception e) {
				logger.error(e.getMessage());
			}
		}
		return min;
	}

}

File: validation/src/main/lua/attributecharacterdistribution.lua
--[[

 This file is part of RoCaWeb.

 RoCaWeb is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>

--]]



-- Implements the KHI-2 validation
-- @author Djibrilla Amadou Kountché 
-- @version 1.0

-- TODO note that this model needs a representative data set otherwise the icd construction will fail

require 'cephes'


local default_alpha = 0.05
local icd = {}
local msg = ""
local expected ={}
local observed = {}
local length = 0



local function printTable(observed)
  local message = "["
  for k,v in ipairs(observed) do
      message = message .. v ..","
  end
  return message .."]"
end


---
-- Sum the values of the table given in parameter between i and j.
--
local function sum(observed, i, j)
   local value = 0

   if observed then
     while observed[i] and i <=j do
        value = value + observed[i]
        i = i+1
     end
   end
   return value
end



-- Create 6 bins for the observed values as  given in Kruegel et al.
local function getBins(observed, expected)
   local values = observed
   local modifiedObserved ={}
   local k = #observed
   local v = #expected
   if k == v then
      return observed
   else
       modifiedObserved[1] = observed[1]
       modifiedObserved[2] = sum(observed, 2, 4)
       modifiedObserved[3] = sum(observed, 5, 7)
       modifiedObserved[4] = sum(observed, 8, 12)
       modifiedObserved[5] = sum(observed, 13, 16)
       modifiedObserved[6] = sum(observed, 17, #observed)

      return modifiedObserved
   end

end


---
--  Khi2 determines the p-value of the
--  return the value of : kh2, distribution value, p-value
local function khi2(observed, expected)
   local khi = 0
   if   #observed ~= #expected then
      -- Log as error
      m.log(1,print('Table size are not equal ' .. #observed .. ' ' .. #expected))
      m.log(1, "Not enough value to do the learning")
   end

   for i = 1, #observed do
      khi = khi + (math.pow((observed[i] - expected[i]),2)/expected[i])
   end
    --Log as debug
    m.log(3,print("khi: " .. khi))
  -- The complementary cumulative distribution

  local pvalue = cephes.chdtrc(#observed-1, khi)

  return khi, cephes.chdtri(#observed - 1, default_alpha), pvalue
end


-- Determine the occurrences of the characters
local function characterOccurences(chaine)
    local pairValues  = {}
    local observed = {}

    -- Count the occurrence of each character
    for i = 1, #chaine do
       local char = chaine:sub(i,i)
       if not pairValues[char]  then
             pairValues[char] = 1
       else
          pairValues[char] = pairValues[char] + 1
       end
     end

     -- Get the occurrence count and sort them
     local i = 1
     for k,v in pairs(pairValues) do
         observed[i] = v
         i = i+1
     end

    table.sort(observed, function(a,b) return a > b end)

    return observed
end



-- Get the character distribution from the rule
local function getICD()
  local expected = {}

  for j=1, 6 do
    local val = m.getvar("TX.icd"..j-1)
    if (val) then expected[j] = tonumber(val) end
  end

  return expected
end


local function initialise()
   length = string.len(params[i].value)
   icd = getICD()
   expected = createExpected(icd, length)


  -- Get the observed values and then compute 6 bins
   observed = characterOccurences(params[i].value)

  observed = getBins(observed,expected)
  if not observed or not expected  then
    m.log(1, "Not enough data to do the validation !")
  end
end


local function validate(param)

  local khi, dist, p_value = khi2(observed,expected)

  if (khi > dist) then
     m.log(1, "\nkhi= "..khi .."\ndist= "..dist .."\n p-value ="..p_value .."\nlen = "..string.len(params[i].value)..  "\n icd= " .. printTable(getICD()) .. "\nobserved=  " .. printTable(observed) .. "\n expected : " .. printTable(expected))
     msg = msg .. param.name .." " .. param.value
     m.log(3, param.name..','..param.value..','..'1')
  else
     m.log(3, param.name..','..param.value..','..'0')
     msg = nil
  end
end

---
-- Create the expected array with is the icd times
-- the length of the current string
local function createExpected(icd, length)
  local expected ={}
   -- Determines the expected values
   for k,v in ipairs(icd) do
      expected[k] = v*length
    end

    return expected
end

-- ModSecurity entry point
function main()

   local msg="";
   local params = m.getvars("ARGS",  { "lowercase", "htmlEntityDecode" })
   local name = m.getvar("TX.name")


    for i = 1,#params do

      if(params[i].name == "ARGS:"..name) then
          validate(param[i])
          break
    end
    end
    return msg
end
